perception/map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -610,17 +615,33 @@ PredictedObject MapBasedPredictionNode::getPredictedObjectAsCrosswalkUser(
   return predicted_object;
 }
 
-double MapBasedPredictionNode::getObjectYaw(const TrackedObject & object)
+void MapBasedPredictionNode::updateObjectData(TrackedObject & object)
 {
-  if (object.kinematics.orientation_availability) {
-    return tf2::getYaw(object.kinematics.pose_with_covariance.pose.orientation);
+  if (
+    object.kinematics.orientation_availability ==
+    autoware_auto_perception_msgs::msg::DetectedObjectKinematics::AVAILABLE) {
+    return;
   }
 
+  // Compute yaw angle from the velocity and position of the object
+  const auto original_object_yaw =
+    tf2::getYaw(object.kinematics.pose_with_covariance.pose.orientation);
   const auto & object_pose = object.kinematics.pose_with_covariance.pose;
   const auto & object_twist = object.kinematics.twist_with_covariance.twist;
   const auto future_object_pose = tier4_autoware_utils::calcOffsetPose(
     object_pose, object_twist.linear.x * 0.1, object_twist.linear.y * 0.1, 0.0);
-  return tier4_autoware_utils::calcAzimuthAngle(object_pose.position, future_object_pose.position);
+  const auto updated_object_yaw =
+    tier4_autoware_utils::calcAzimuthAngle(object_pose.position, future_object_pose.position);
+
+  object.kinematics.pose_with_covariance.pose.orientation =
+    tier4_autoware_utils::createQuaternionFromYaw(updated_object_yaw);
+
+  // If the object yaw flips, we also flip the velocity
+  if (std::fabs(updated_object_yaw - original_object_yaw) > tier4_autoware_utils::pi / 2.0) {
```

Comment:
@purewater0901 
calcAzimuthAngle returns -180deg ~180deg using atan2 is it OK without normalizing here?
I think it's necessary